presto-main/src/main/java/io/prestosql/sql/planner/DynamicFilter.java

```diff
@@ -0,0 +1,261 @@
+            }
+            else {
+                break;
+            }
+        }
+        log.warn("createDynamicFilter: unsupported node %s", probeNode);
+        return Optional.empty();
+    }
+
+    public static Optional<DynamicFilter> create(JoinNode node, PlanNode probeNode, List<Symbol> probeSymbols, PlanNode buildNode, List<Symbol> buildSymbols, TypeProvider types)
+    {
+        if (node.getType() != JoinNode.Type.INNER) {
+            log.debug("createDynamicFilter: dynamic filters are supported only for inner-joins");
+            return Optional.empty();
+        }
+        if (!node.getDistributionType().equals(Optional.of(JoinNode.DistributionType.REPLICATED))) {
+            log.debug("createDynamicFilter: dynamic filters are supported only for replicated joins");
+            return Optional.empty();
+        }
+
+        int numOfJoinSymbols = probeSymbols.size();
+        if (numOfJoinSymbols != buildSymbols.size()) {
+            // TODO: consider matching some of the symbols for partial dynamic filtering
+            log.debug("createDynamicFilter: buildSymbols' size (%d) doesn't match probeSymbols' size (%d)",
+                      numOfJoinSymbols, probeSymbols.size());
+            return Optional.empty();
+        }
+
+        return findTableScanNode(probeNode, probeSymbols).map(scanNode -> {
+            ImmutableList.Builder<DynamicFilter.Descriptor> dynamicFilters = new ImmutableList.Builder<>();
+            for (int i = 0; i < numOfJoinSymbols; ++i) {
+                Symbol probeSymbol = probeSymbols.get(i);
+                Symbol buildSymbol = buildSymbols.get(i);
+
+                Type buildType = types.get(buildSymbol);
+                int buildChannel = buildNode.getOutputSymbols().indexOf(buildSymbol);
+
+                ColumnHandle probeColumn = scanNode.getAssignments().get(probeSymbol);
+                dynamicFilters.add(new DynamicFilter.Descriptor(buildType, buildChannel, probeColumn));
+            }
+            log.debug("createDynamicFilter: dynamic filters initialized for join node: %s", node);
+            return new DynamicFilter(dynamicFilters.build());
+        });
+    }
+
+    public class Collector
+    {
+        private final ImmutableList.Builder<Page> pageBuilder;
+        private final PageChannelSelector selector;
+        private int positionsCount;
+        private long sizeInBytes;
+
+        public Collector()
+        {
+            this.pageBuilder = new ImmutableList.Builder<>();
+            int[] buildChannels = descriptors.stream().mapToInt(d -> d.buildChannel).toArray();
+            this.selector = new PageChannelSelector(buildChannels);
+            this.positionsCount = 0;
+            this.sizeInBytes = 0L;
+        }
+
+        public void add(Page page)
+        {
+            // Collect only the columns which are relevant for the JOIN.
+            Page projectedPage = selector.apply(page);
+            positionsCount += projectedPage.getPositionCount();
```

Comment:
We would want to count only non-null values in `positionsCount` and `sizeInBytes` as null values won't contribute to creating a large predicate. I don't think current code takes care of that case.

Reply:
Good idea, thanks!
Please take a look at https://github.com/rzeyde-varada/prestosql/commit/edb2d36b90517b5601136e56c8f11ef835a56e97 (pushed in a separate [wip/dynamic-filter](https://github.com/rzeyde-varada/prestosql/commits/wip/dynamic-filter) branch).

A small question about `Block#getEstimatedDataSizeForStats` - I have used this method for estimating the byte size of each value in order to compute `sizeInBytes` for dynamic filter collection. Is it OK?

Reply:
I think it should be ok, an alternative is to use `value.toString().getBytes().length`, though it looks like `getEstimatedDataSizeForStats` might be the more straightforward way.
You might want to check the performance impact of using HashSet to get distinct values count in rzeyde-varada@edb2d36. If the overhead is minimal then getting distinct values is good, we had avoided it earlier to keep this part as lightweight as possible.

Reply:
> You might want to check the performance impact of using HashSet to get distinct values count in [rzeyde-varada/prestosql@edb2d36](https://github.com/rzeyde-varada/prestosql/commit/edb2d36). 

Thanks, will add a benchmark.

Reply:
Added a simple benchmark, rebased and force-pushed 733ffaf10c.
When running on my machine, the overhead seems to be quite low:

With `collectOperator` running after `tableScanOperator`:
```
dynamic_filter_collect ::   20.360 cpu ms ::    0B peak memory :: in   15K,      0B,    737K/s,      0B/s :: out   15K,   132KB,    737K/s,  6.32MB/s
```

Without `collectOperator` (only `tableScanOperator`):
```
dynamic_filter_collect ::   19.855 cpu ms ::    0B peak memory :: in   15K,      0B,    755K/s,      0B/s :: out   15K,   132KB,    755K/s,  6.48MB/s
```

Reply:
LGTM

Reply:
Thanks, will cherry-pick https://github.com/prestosql/presto/commit/733ffaf10ccd86bc1872d90ed54d2440c3fbad2b into `dynamic-filter` branch.